CompositeImplicitAutograd compliance testing

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#65819 CompositeImplicitAutograd compliance testing**

Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

zou3519 committed:
CompositeImplicitAutograd compliance testing

Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+            unwrapped_rs = func(*unwrapped_args, **unwrapped_kwargs)
+            rs = tree_map(wrap, unwrapped_rs)
+
+        if is_view_fn(func):
```
Is there a way to check if there's any storage aliasing going on, without having to check if the op is a view op?

E.g. if you have these for pieces of data `args`, `unwrapped_args`, `unwrapped_results`, `results`, then maybe you could do something like:
```
# ... first ensure that args[0] and result[0] are both valid tensors, since all view ops have a first arg with tensor type
if unwrapped_args[0].storage().is_alias_of(unwrapped_results[0].storage()):
```

I'm not 100% that all of those API's are exposed in python though

zou3519 replied:
That's a good point, thanks for the suggestion. There are some APIs that are exposed; e.g. we can do
```
unwrapped_results[0]._base is unwrapped_args[0]
```
Let me go play around with this

zou3519 replied:
Hmm, I was unable to get the following to work:
- the `unwrapped_results[0]._base is unwrapped_args[0]` doesn't work because torch_dispatch runs after autograd, so the _base field never gets populated (that happens somehow during the autograd logic?)
- binding `is_alias_of` into Python doesn't work, because more than just view functions return aliases (in-place operations is one example). The later code segfaults if presented with an in-place operation, which is not good.

So the conclusion here is: we probably want some introspection at the python level to be able to tell if something is a view operation?

albanD replied:
> that happens somehow during the autograd logic?

Yes this is part of the InplaceOrView autograd key.

> Is there a way to check if there's any storage aliasing going on, without having to check if the op is a view op?

We actually don't (should not) have multiple storages that look at the same memory. And if the input and output share their storage, they have to alias (we don't have out= variants here). So something like `unwrapped_args[0].storage().data_ptr() == unwrapped_results[0].data_ptr()` would be enough no?

Also because data_ptr() is bad, we could fix the storage python binding to have `unwrapped_args[0].storage() is unwrapped_results[0].storage()` work :D

zou3519 replied:
> unwrapped_args[0].storage().data_ptr() == unwrapped_results[0].data_ptr() would be enough no?

It turns out this doesn't work for tensors with 0-size dimensions. The data_ptr() returns 0

> Also because data_ptr() is bad, we could fix the storage python binding to have unwrapped_args[0].storage() is unwrapped_results[0].storage() work :D

Yes, that would be awesome (going to file an issue...)

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+        '_conj',
+    }
+
+def is_inplace_view_fn(func):
```
assuming that you can do away with the hardcoded view list and just need to deal with view_inplace ops: I need this list for functionalization too, and I'm doing it through tagging in native_functions.yaml. At some point that'll probably be available for introspection (probably not before this PR lands though).

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+        # The storage of CompositeCompliantTensor should never be used directly
+        # by a CompositeImplicitAutograd operation; if the CompositeImplicitAutograd
+        # operator attempts to read from the storage without dispatching then it'll
+        # raise a RuntimeError due to it being a meta storage.
```
will this raise whatever default error gets thrown by `storage_access_should_throw`? We probably want this test to give a more actionable error message.

zou3519 replied:
The test catches the error, adds some context with actionable information, and then rethrows the error with additional context. So it looks like the following:
```
================================================= FAILURES =================================================
___________________ TestCommonCPU.test_composite_compliance_linalg_cholesky_cpu_float32 ____________________
Traceback (most recent call last):
  File "/raid/rzou/pt/workspace/torch/testing/_internal/composite_compliance.py", line 165, in _check_compos
ite_compliance
    op(*args, **kwargs)
  File "/raid/rzou/pt/workspace/torch/testing/_internal/common_methods_invocations.py", line 627, in __call_
_
    return self.op(*args, **kwargs)
  File "/raid/rzou/pt/workspace/torch/_tensor.py", line 1064, in __torch_function__
    ret = func(*args, **kwargs)
RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a
 lazy allocation, so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/raid/rzou/pt/workspace/test/test_ops.py", line 515, in test_composite_compliance
    _check_composite_compliance(op, args, kwargs)
  File "/raid/rzou/pt/workspace/torch/testing/_internal/composite_compliance.py", line 167, in _check_compos
ite_compliance
    raise RuntimeError(f"CompositeImplicitAutograd compilance check failed with "
RuntimeError: CompositeImplicitAutograd compilance check failed with the following error. If you are adding
an OpInfo of an existing operator, please feel free to skip this test because the problem was pre-existing a
nd file an issue. Otherwise, if you added a new operator, please read through the CompositeImplicitAutograd
Compliance section in aten/src/ATen/native/README.md for how to resolve this. Got error message: The tensor
has a non-zero number of elements, but its data is not allocated yet. Caffe2 uses a lazy allocation, so you
will need to call mutable_data() or raw_mutable_data() to actually allocate memory.

```

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
```
I have a nagging feeling that this list of metadata is incomplete, but it's probably easy enough to add to over time

zou3519 replied:
Yeah, it's definitely not complete and we can add more over time.

ezyang replied:
Isn't this simply a matter of going over all the fields in TensorImpl? This is decidedly finite.

zou3519 replied:
Went over TensorImpl fields and added the ones that were exposed in Python, with some exception... (see other comment)

albanD commented on aten/src/ATen/native/README.md:
```diff
@@ -346,6 +346,36 @@ added if applicable), so that it's still available for other backends to use.
+that is not CompositeImplicitAutograd.
+
+CompositeImplicitAutograd operators must not:
+- call `set_` or `resize_` or moral equivalents. These are tricky to handle for
```
I'm not sure why set_ is here? It is going through the dispatcher and is not more problematic than any other inplace op?
Also it is used by python code quite a bit (nn.Module in particular) so I don't think we want to "encourage" backend writers not to support it.

zou3519 replied:
We can remove `set_`. I'm just concerned people will use `set_` as a replacement for `resize_` when writing kernels

zou3519 replied:
@albanD I ended up removing `set_` from the list.

albanD replied:
ok!

albanD commented on aten/src/ATen/native/README.md:
```diff
@@ -346,6 +346,36 @@ added if applicable), so that it's still available for other backends to use.
+- Change the metadata of a Tensor without performing dispatches. Examples of these
+operations are directly accessing the TensorImpl API to modify the
+sizes/strides/metadata of a Tensor.
+- In the same vein as the last point, `data_ptr` access or `item` access are not
```
item should be fine no? It gets a plain number as output

zou3519 replied:
item isn't fine because meta tensors and wrapper Tensors choke on it

zou3519 replied:
To clarify a bit more... in C++ item() is literally just a call to .data_ptr and doesn't go through the dispatcher.

albanD replied:
ok

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+                f"{func.__name__} is not allowed to be called inside of "
+                f"CompositeImplicitAutograd operators.")
+
+        with no_dispatch():
```
This is not needed right?

zou3519 replied:
This is needed to prevent infinite recursion if `func` is a factory function

albanD replied:
We should have a better story for this. In particular, this prevents any interaction between different subclasses because other subclass won't have their dispatch called anymore...

albanD replied:
btw I was just chatting with Anjali where the disable torch function was introducing subtle bugs in her implementation when subclass are involved.
For these, I think `super().__torch_function__(...)` is really the right way to go. I feel like we need a similar thing for torch_dispatch no?

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+            # are the same. Here we try to make B alias A to avoid those asserts.
+            aliased_tensor = args[0]
+            with no_dispatch():
+                x = torch.empty(aliased_tensor.shape, dtype=aliased_tensor.dtype,
```
You don't need to create this one full size. You can pass an empty tuple as the size.

zou3519 replied:
Good point!

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+            # Autograd asserts that for B = A.view_fn(...), B and A's storages
+            # are the same. Here we try to make B alias A to avoid those asserts.
+            aliased_tensor = args[0]
+            with no_dispatch():
```
not needed either?

zou3519 replied:
Unfortunately we need this to prevent infinite recursion on torch.empty

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+
+        if is_view_fn(func):
+            # Autograd asserts that for B = A.view_fn(...), B and A's storages
+            # are the same. Here we try to make B alias A to avoid those asserts.
```
Could you explain how this call is different from the one that already happened above?

zou3519 replied:
The above call happens on the unwrapped tensors. The call below happens on the wrapped tensors. This is a really weird way of getting the wrappers' storage to alias each other: the no_dispatch means that we completely ignore the fact that the input wrapper tensor has a PyObj associated with it, we pretend it is just a regular meta tensor, and then the operation produces a meta tensor output that aliases the input.

zou3519 replied:
Added a comment into the code that better explains this. Please let me know if that makes sense

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+        # run into these, we manually modify the metadata of the input.
+        with no_dispatch():
+            if is_inplace_view_fn(func):
+                func(args[0])
```
these func might have args/kwargs no?

zou3519 replied:
Good point, thanks for the catch!

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,174 @@
+    args = tree_map(wrap, args)
+    kwargs = tree_map(wrap, kwargs)
+    try:
+        with enable_python_mode(CompositeCompliantTensor):
```
Why do you need this to be a mode?

zou3519 replied:
We would like to catch CompositeImplicitAutograd implementations where someone creates an empty tensor and then calls `.resize_` on it. The empty tensor is not an input of the function but the mode makes it so that it gets wrapped in the CompositeCompliantTensor

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+            # Autograd asserts that for B = A.view_fn(...), B and A's storages
+            # are the same. Here we try to make B alias A to avoid those asserts.
+            with no_dispatch():
+                # Idea: this is a weird way of getting a storage that aliases the input.
+                # 1. under no_dispatch, all of the wrapper tensors look like meta tensors
+                # 2. we run func, which ends up running the meta tensor version of the view op
+                # 3. the result is one or more meta tensor that aliases the input.
+                # 4. we set the storage of the wrapper tensor results to be
+                #    the meta tensor(s) that aliases the input
+                result = func(*args, **kwargs)
+                if isinstance(result, tuple) or isinstance(result, list):
+                    for a, b in zip(rs, result):
+                        a.set_(b)
+                else:
+                    rs.set_(result)
```
@albanD, @bdhirsh: if we think the "make the wrapper output's storage alias the wrapper input's storage" logic is too hacky and if we're willing to accept that these tests do not pass in DEBUG mode while we figure out a better solution (most of `test_python_dispatch` fails under DEBUG mode) then I can delete this code for now

ezyang replied:
To me, this seems like the starting point for a proper fix to #65339. So... it's OK for now, let's just keep making it better?

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -346,6 +346,36 @@ added if applicable), so that it's still available for other backends to use.
+allowed. These operations do not go through the dispatcher.
+- `copy_` is a marginal case. If you're able to rewrite your operation without
+`copy_` you should definitely do so; this should be trivial if you're not copy-ing
+into a view. Otherwise, it is fine to leave the code as-is.
```
My kingdom for a static analysis instead. cc @mruberry

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+from functools import partial
+from torch.utils._python_dispatch import enable_python_mode
+
+# TODO: move this into library proper
```
NOPE you get to fix this TODO 🤪

ezyang replied:
torch/overrides.py is reasonable; otherwise we should make a new toplevel for dispatch-y related functionality

zou3519 replied:
Lol this is the 2nd time I've copied this function

albanD replied:
Do we really want to make this a thing? It doesn't feel like the right fix for this problem.
If people start to use this, then all the nice use case from https://colab.research.google.com/drive/12FN24CPRe_FVxjE46SZTu8q3j4RVSq15#scrollTo=XVeRMjCT8ptG would break because the other subclass would never be called...

ezyang replied:
So @zou3519 talked a while back about a more surgical variant of this that simply removes a class from consideration. I'm not sure this is still the right thing to do. In general, some tensors will want a way to say "treat me like my superclass without unwrapping"; where a super() call is inappropriate.

albanD replied:
Actually I was just debugging something with Anjali related to torch_function and using disable torch function that was creating subtle issues when subclassing.
In this case, proper use of super() is doing the right thing because we go all the way to the `Tensor.__torch_function__` when everything else has been handled and we're good to go.
We really need a mechanism to be able to do the same in torch_dispatch no?

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+        '_unsafe_view',
+        '_conj',
+        'alias',
+    }
```
My kingdom for some sort of introspection mechanism on the operators...

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+            return e.elem if isinstance(e, CompositeCompliantTensor) else e
+
+        def wrap(e):
+            return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
```
A fridge thought: shouldn't pytree only ever call wrap on Tensors...

zou3519 replied:
pytree APIs (like tree_map) are useful for non-tensors. For example, `vmap(f, in_dims)(*args)`: the `in_dims` is allowed to be a pytree of the same "structure" as args and what we want to do there is "zip" the in_dims with the `args`.

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+                           f"Otherwise, if you added a new operator, please read "
+                           f"through the CompositeImplicitAutograd Compliance section in "
+                           f"aten/src/ATen/native/README.md for how to resolve this. "
+                           f"Got error message: {err.args[0]}")
```
If you want to chain, you can just say `raise ... from err`

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+                # 2. we run func, which ends up running the meta tensor version of the view op
+                # 3. the result is one or more meta tensor that aliases the input.
+                # 4. we set the storage of the wrapper tensor results to be
+                #    the meta tensor(s) that aliases the input
```
Can we quote https://github.com/pytorch/pytorch/issues/65339 here

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,172 @@
+                result = func(*args, **kwargs)
+                if isinstance(result, tuple) or isinstance(result, list):
+                    for a, b in zip(rs, result):
+                        a.set_(b)
```
Do you get correct storage offsets this way too? Seems like they wouldn't propagate correctly

ezyang reviewed (approved):
okey dokey

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,188 @@
+def check_metadata_consistency(wrapper_tensor):
+    if not isinstance(wrapper_tensor, CompositeCompliantTensor):
+        return
+    things_to_check = {
+        'shape': Tensor.size,
+        'dtype': lambda x: x.dtype,
+        'device': lambda x: x.device,
+        'numel': Tensor.numel
+    }
```
@albanD @ezyang: It turns out that a wrapper tensor does not have the same strides or storage_offset as its element. Should it?

I can see this not being required for wrapper tensors like SparseTensor where strides/storage_offset make no sense. However this may be useful for something like functorch's PythonTensor that does tracing: if we trace a concrete tensor, then the wrapper should have the same strides/storage offset.

ezyang replied:
It should.

zou3519 replied:
Thanks for the feedback. Change to wrapper object creation coming soon...

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,191 @@
+        'squeeze_',
+        'unsqueeze_',
+        'transpose_',
+        't_',
```
Does this need to contain every inplace_view op? I know of a couple others, like `swapdims_`, that technical fit the description. Although maybe the main difference is that `swapdims_` is CompositeImplicitAutograd and the ones in this list are not.

bdhirsh replied:
(I quickly checked by comparing with the `inplace_view` tag in `native_functions.yaml`)

bdhirsh replied:
oh yeah I guess the whole point is that this is used inside of `__torch_dispatch__`, meaning that it will never be run on a `CompositeImplicitAutograd` op anyway (because the composite op will have been decomposed before we hit `__torch_dispatch__`)

bdhirsh replied:
Although maybe we should just add all of them in for consistency anyway, since I see some composite view ops in the view list above (like `movedim`). At least we're being overly conservative, so that we still get coverage if someone happens to change a composite view op to be non-composite.

zou3519 replied:
We only need the ones that are not CompositeImplicitAutograd, but I'll add all of them in for consistency

bdhirsh commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -0,0 +1,191 @@
+                # 4. we set the storage (and sizes/strides/offset) of the wrapper
+                #    tensor results to be that of the tensors that alias the input
+                result = func(*args, **kwargs)
+                if isinstance(result, tuple) or isinstance(result, list):
```
nit: should/can this be done with `tree_map`?

bdhirsh reviewed (approved):
Left minor comments but LGTM! This will also serve as a nice template PR for adding new OpInfo functionality (like functionalization testing)

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).

zou3519 committed:
Update on "CompositeImplicitAutograd compliance testing"


Related to #61669.

Functions registered as CompositeImplicitAutograd MUST work for most, if
not all, backends. This includes Tensor subclasses.

To achieve this, we (PyTorch) impose a set of constraints on how a
CompositeImplicitAutograd function can be written.

Concretely, this PR adds tests for all OpInfos that checks for
compliance. The things that get tested in this PR apply to composite
ops and are that:
- the op does not change the metadata of a Tensor without performing
dispatches
- the op does not call set_ or resize_
- the op does not directly access the data ptr

The mechanism for the test is to create a new __torch_dispatch__
object, CompositeCompliantTensor. For each operator, we wrap all inputs
in CompositeCompliantTensor, turn on python mode for it,
and send it through the operator.

Non-CompositeImplicitAutograd operators will pass the test because they
perform a dispatch to backend code. Here's how CompositeCompliantTensor
catches problems:

- If it sees set_ or resize_ getting called, it will directly error
out
- After each operation, CompositeCompliantTensor checks to make sure
that its metadata is consistent with that of the thing it is wrapping.
If the CompositeImplicitAutograd op modifes the metadata directly
(through e.g. the TensorImpl API) then the metadata will go out of sync.
- If data_ptr gets called, that returns a nice error (because the
storage is meta).

CompositeCompliantTensor is written in an interesting way. First off,
if a view operation occurs (e.g. `B = A.view_op(...)`), then B.storage()
must alias A.storage() where B.storage() is CompositeCompliantTensor's
storage, NOT the storage of the tensor it is wrapping. This is an
invariant in autograd, see #62182 for details. To handle
this we replay the view on A's storage and set it as B's storage.

Secondly, there are cases where the metadata is allowed to go out of
sync. I believe this is only possible with in-place view functions, like
transpose_, t_, squeeze_, unsqueeze_. Those are special cased.

Finally, I added a new section to aten/src/ATen/native/README.md about
what it means to be CompositeImplicitAutograd Compliant

Test Plan:
- run tests

Differential Revision: [D31268369](https://our.internmc.facebook.com/intern/diff/D31268369)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31268369).
